burotel/indico_burotel/util.py

```diff
@@ -0,0 +1,119 @@
+# the LICENSE file for more details.
+
+from __future__ import unicode_literals
+
+import itertools
+
+from dateutil.relativedelta import relativedelta
+from dateutil.rrule import rrule, MONTHLY
+from sqlalchemy.orm import aliased
+from sqlalchemy.sql.expression import bindparam
+
+from indico.core.db import db
+from indico.modules.rb.models.rooms import Room
+from indico.modules.rb.models.reservations import Reservation, ReservationOccurrence
+
+
+def _build_per_building_query(*query_results):
+    return (db.session
+            .query(*query_results)
+            .select_from(ReservationOccurrence)
+            .filter(Room.is_reservable, ~Room.is_deleted)
+            .filter(ReservationOccurrence.is_valid, Reservation.is_accepted)
+            .join(Reservation)
+            .join(Room)
+            .group_by(Room.building, Room.division))
+
+
+def calculate_monthly_stats(start_dt, end_dt):
+    """Calculate monthly stats for the Burotel system, based on a date range."""
+
+    room = aliased(Room)
+    months = list(rrule(freq=MONTHLY, dtstart=start_dt, until=end_dt))
+
+    desk_count = (db.session.query(db.func.count(room.id))
+                    .filter(
+                        Room.building == room.building,
+                        Room.division == room.division,
+                        room.is_reservable,
+                        ~room.is_deleted)
+                  ).label('desk_count')
+
+    # a first query which retrieves building data as well as the total number of bookings
+    building_query = _build_per_building_query(
+        Room.building.label('number'),
+        Room.division.label('experiment'),
+        desk_count,
+        db.func.count(
+            db.func.concat(Reservation.id, ReservationOccurrence.start_dt)).label('bookings')
+        ).filter(ReservationOccurrence.start_dt >= start_dt, ReservationOccurrence.end_dt <= end_dt)
+
+    parts = []
+    for n, month_start in enumerate(months):
+        month_end = (month_start + relativedelta(months=1, days=-1)).replace(hour=23, minute=59)
+        parts.append(
+            _build_per_building_query(
+                Room.building.label('number'),
+                Room.division.label('experiment'),
+                bindparam('month-{}'.format(n), n).label('month'),
+                db.func.count(db.func.concat(Reservation.id, ReservationOccurrence.start_dt)).label('bookings')
+            ).filter(ReservationOccurrence.start_dt >= month_start, ReservationOccurrence.end_dt <= month_end))
+
+    # create a union with all month queries. this will return a (second) query which will provide
+    # separate totals for each month
+    month_query = parts[0].union(*parts[1:])
+
+    # group the results of the "building query" by building number/experiment
+    bldg_exp_map = []
+    for k, v in itertools.groupby(
+            sorted(building_query, key=lambda b: (b.number, b.experiment)),
+            lambda b: (b.number, b.experiment)):
+        values = list(v)
```

Comment:
we have a util somewhere in the indico core that does the groupby with sorting and materializing lists

Reply:
I actually didn't need that `group_by` 🤷‍♂